RE-1325 Create pluggable artifact upload hook

odyssey4me opened the pull request:
In this PR, we create a pluggable artifact upload hook which currently implements two types: 'log' and 'file'.

This new hook makes use of the in-repo ``component_metadata.yml`` file to discover its configuration.

The 'log' type works in exactly the same way as the current job artifact publishing, and has a shim for the ``RE_HOOK_ARTIFACT_DIR`` environment variable for backwards compatibility.

The 'file' type provides the capability to upload any file or recursively upload any folder to public cloud's cloudfiles, in a container name which matches the repository running the test.

We also switch the current log archiving mechanism to use the 'log' type.

There is follow up work in [RE-1813](https://rpc-openstack.atlassian.net/browse/RE-1813) to automatically publish the URL where these files can be accessed to some place that's easy to find.

Issue: [RE-1325](https://rpc-openstack.atlassian.net/browse/RE-1325)
Issue: [RE-1812](https://rpc-openstack.atlassian.net/browse/RE-1812)

odyssey4me commented:
recheck_issue_link

hughsaunders commented on docs/debugging.md:
```diff
@@ -25,13 +25,50 @@ it's inefficient to wade through it when looking for the cause of a failure.
+
+To view ``file`` artifacts, navigate to ``<repo name>.re.rpc.rackspace.com``
+(eg: http://rpc-octavia.re.rpc.rackspace.com). If this is not available, please
+contact the RE team to create the convenience alias in DNS for the git
```
Is there a procedure for this somewhere?

odyssey4me replied:
Not currently. I will be adding it to confluence once this is accepted, and once I have a sample repo to figure out the procedure.

odyssey4me commented on playbooks/artifact_upload.yml:
```diff
@@ -0,0 +1,54 @@
+          {%-   endif %}
+          {%- else %}
+          {%-   set _result = legacy_artifact_log_type %}
+          {%- endif %}
```
I tried to keep the specifics (uncluding checking if the source exists) of each type in the task files for that type, rather than here. It may be slightly less efficient, but it makes it more obvious why an artifact did not get uploaded in the resulting console output because you're able to see that the type config was picked up (or not), but that it got skipped because the source did not exist. I figure that this is more useful for diagnosing issues than skipping it earlier where the cause of skipping would be less obvious.

odyssey4me commented:
@hughsaunders The 'file' type artifact is something of a work-in-progress, whereas the 'log' type is an adaptation of the well established existing process (and is tested by the unit tests).

I've been thinking today that perhaps instead of using a convenience CNAME in DNS (which is a manual task by a restricted set of people), we could automatically push up a PR to the release repo to add the CDN URL's once an artifact has been published. This would provide much more transparency and do away with the need for the CNAME.

Regardless, the primary intent of this card is to establish something to publish artifacts more easily. I'd like to get on with the Epic's intent (to publish MNAIO images), then perhaps circle back in the same epic to make the 'file' type publishing more visible and robust.

odyssey4me commented:
recheck_all

odyssey4me commented:
Successful build with new job configuration:
https://rpc.jenkins.cit.rackspace.net/job/RE-unit-test-artefact-publication-odyssey4me/6
CDN Container public URL for rpc-gating: https://3532b781e8c3821b0e1d-056fbf9cb875ed1243d1c670da7bbdc4.ssl.cf1.rackcdn.com/

hughsaunders commented on playbooks/tasks/artifact_upload/upload_file.yml:
```diff
@@ -0,0 +1,67 @@
+---
+- name: Check that the artifacts source path has been provides
```
provided

hughsaunders commented on playbooks/tasks/artifact_upload/upload_file.yml:
```diff
@@ -0,0 +1,67 @@
+      command: >-
+        swift upload {{ object_store_container_name }} {{ artifact['source'] | basename }}
+        --object-threads 100
+        {{ artifact.expire_after is defined | ternary("--header 'X-Delete-After:" ~ artifact.get('expire_after') ~ "'", "") }}
```
why `~` instead of `+` ?

odyssey4me replied:
``~`` is for concatenating strings. See http://jinja.pocoo.org/docs/2.10/templates/#math

odyssey4me commented on playbooks/artifact_upload.yml:
```diff
@@ -0,0 +1,73 @@
+  vars:
+    region: "DFW"
+    cloud_name: "public_cloud"
+    component_metadata_file_path: "{{ lookup('env', 'WORKSPACE') }}/component_metadata.yml"
```
It probably should. I know we have ``RE_JOB_REPO_NAME``, but is there perhaps a better way to get the repo name/path?

hughsaunders replied:
I'd use `"${env.WORKSPACE}/${env.RE_JOB_REPO_NAME}"` which is what the clone uses, so you'll definitely end up in the same place..

odyssey4me commented on rpc_jobs/unit/artefact_publish.yml:
```diff
@@ -27,6 +27,22 @@
+              ]
+            ]
+          ]
+          writeYaml file: 'component_metadata.yml', data: artifacts_data
```
Correct, will fix that - well spotted!

mattt416 replied:
Good catch -- and explains why `component_metadata_file_path: "{{ lookup('env', 'WORKSPACE') }}/component_metadata.yml"` is working in `playbooks/artifact_upload.yml`.

Side note, may be worth trying to test this on a live job to suss out any errors like this.

mattt416 commented on docs/debugging.md:
```diff
@@ -23,28 +23,38 @@ operations such as creating a single use instance for a project's jobs to
+"artifacts": [
+  {
+      "type": "log",
+      "source": "/foo"
```
Total nit, but weird indentation here.

odyssey4me replied:
Ah yeah, looks like I used one too many spaces?

mattt416 commented on pipeline_steps/common.groovy:
```diff
@@ -301,7 +301,18 @@ String container_name(){
+      // If the build fails, or this is a PR, the default
+      // artifact types to upload are 'log'. Otherwise we
+      // will attempt all types.
+      if ((currentBuild.result == "FAILURE") || (env.'ghprbPullId') == '') {
```
I don't think this comment matches the actual code here.

odyssey4me replied:
It looks like it does to me?

mattt416 replied:
If it's a PR, wouldn't `env. ghprbPullId != ''` ?

odyssey4me replied:
Ah, great catch!

mattt416 commented on playbooks/artifact_upload.yml:
```diff
@@ -0,0 +1,73 @@
+    legacy_artifact_log_type:
+      - "type": "log"
+        "source": "{{ lookup('env', 'RE_HOOK_ARTIFACT_DIR') | default(lookup('env', 'WORKSPACE') ~ '/artifacts', true) }}"
+    artifact_type_include: "{{ (lookup('env', 'ghprbPullId') == '') | ternary('all', 'log') }}"
```
Can you not use `artifact_type_include` here ?

odyssey4me replied:
Well yes, I could just leave this out and not define it here - expecting it to be passed in. However, as @hughsaunders noted having it here is useful for testing outside of jenkins.

mattt416 replied:
OK, makes sense. Probably no point passing it in from groovy then?

odyssey4me replied:
We pass it in from groovy for two reasons:

1. To allow our PR unit test to be able to upload more than just logs.
2. To allow any pipeline to set the type they want to upload as they wish, much like we want to for the unit test. :)

hughsaunders commented on playbooks/artifact_upload.yml:
```diff
@@ -0,0 +1,73 @@
+    legacy_artifact_log_type:
+      - "type": "log"
+        "source": "{{ lookup('env', 'RE_HOOK_ARTIFACT_DIR') | default(lookup('env', 'WORKSPACE') ~ '/artifacts', true) }}"
+    artifact_type_include: "{{ (lookup('env', 'ghprbPullId') == '') | ternary('all', 'log') }}"
```
That would work, simpler and means we only check the trigger in one place.

hughsaunders commented on pipeline_steps/pubcloud.groovy:
```diff
@@ -234,24 +234,19 @@ def runonpubcloud(Map args=[:], Closure body){
+def uploadArtifacts(Map args = [:]){
+  print("Uploading artifacts.")
+  clouds_cfg = common.writeCloudsCfg()
+  withEnv(["OS_CLIENT_CONFIG_FILE=${clouds_cfg}"]){
```
Ah sorry, I should have spotted it was a moved block, no pressure to update it.

odyssey4me commented:
Updated job result: https://rpc.jenkins.cit.rackspace.net/job/RE-unit-test-artefact-publication-odyssey4me/10
Updated CDN result, in the right destination this time: https://3532b781e8c3821b0e1d-056fbf9cb875ed1243d1c670da7bbdc4.ssl.cf1.rackcdn.com/RE-unit-test-artefact-publication-odyssey4me/10/

mattt416 commented on pipeline_steps/common.groovy:
```diff
@@ -301,7 +301,18 @@ String container_name(){
+      // If the build fails, or this is a PR, the default
+      // artifact types to upload are 'log'. Otherwise we
+      // will attempt all types.
+      if ((currentBuild.result == "FAILURE") || (env.'ghprbPullId') != '') {
```
I think you should actually be checking that `env.'ghprbPullId') != null` ... sorry, didn't catch that the first time around!
